utilities/bin2ass/MIPSsim/MIPSsim.cpp

```diff
@@ -0,0 +1,654 @@
+#include <iostream>
+#include <sstream>
+#include <fstream>
+#include <string>
+#include <vector>
+#include <map>
+#include <utility>
+#include <functional>
+#include <bitset>
+#include <cmath>
+
+using namespace std;
+
+inline string operator "" _s(const char* str, size_t /*length*/)
+{ 
+    return string(str); 
+}
+
+int32_t twosComp2Decimal(string twos_comp32, bool is_signed=false);
+
+class MembershipChecker
+{
+  map<string, function<bool(string)>> func_map;
+
+public:
+  MembershipChecker()
+  {
+    func_map["CAT1"] = [this](string instruction){ return isCAT1(instruction); };
+    func_map["CAT2"] = [this](string instruction){ return isCAT2(instruction); };
+    func_map["CAT3"] = [this](string instruction){ return isCAT3(instruction); };
+
+    func_map["NOP"] = [this](string instruction){ return isNOP(instruction); };
+    func_map["J"] = [this](string instruction){ return isJ(instruction); };
+    func_map["BEQ"] = [this](string instruction){ return isBEQ(instruction); };
+    func_map["BNE"] = [this](string instruction){ return isBNE(instruction); };
+    func_map["BGTZ"] = [this](string instruction){ return isBGTZ(instruction); };
+    func_map["SW"] = [this](string instruction){ return isSW(instruction); };
+    func_map["LW"] = [this](string instruction){ return isLW(instruction); };
+    func_map["BREAK"] = [this](string instruction){ return isBREAK(instruction); };
+
+    func_map["XOR"] = [this](string instruction){ return isXOR(instruction); };
+    func_map["MUL"] = [this](string instruction){ return isMUL(instruction); };
+    func_map["ADD"] = [this](string instruction){ return isADD(instruction); };
+    func_map["SUB"] = [this](string instruction){ return isSUB(instruction); };
+    func_map["AND"] = [this](string instruction){ return isAND(instruction); };
+    func_map["OR"] = [this](string instruction){ return isOR(instruction); };
+    func_map["ADDU"] = [this](string instruction){ return isADDU(instruction); };
+    func_map["SUBU"] = [this](string instruction){ return isSUBU(instruction); };
+
+    func_map["ORI"] = [this](string instruction){ return isORI(instruction); };
+    func_map["XORI"] = [this](string instruction){ return isXORI(instruction); };
+    func_map["ADDI"] = [this](string instruction){ return isADDI(instruction); };
+    func_map["SUBI"] = [this](string instruction){ return isSUBI(instruction); };
+    func_map["ANDI"] = [this](string instruction){ return isANDI(instruction); };
+    func_map["SRL"] = [this](string instruction){ return isSRL(instruction); };
+    func_map["SRA"] = [this](string instruction){ return isSRA(instruction); };
+    func_map["SLL"] = [this](string instruction){ return isSLL(instruction); };
+  }
+
+  bool operator()(string instruction, string instruction_type)
+  {
+    return func_map[instruction_type](instruction);
+  }
+
+  bool isCAT1(string inst)
+  {
+    return inst.substr(0, 3) == "001";
```

Comment:
No, STL writers have implemented these kind of common comparisons with efficiency in mind, its well optimized.